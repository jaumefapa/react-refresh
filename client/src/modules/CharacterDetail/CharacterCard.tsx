import { character } from '../../interfaces/interfaces'
import FavoriteButton from '../common/FavoriteButton/FavoriteButton'
import './characterCard.css'

export default function CharacterCard({ character }: { character: character }) {
  return (
    <div className="character-card-container">
      <img
        className="character-card-image"
        src={character.image}
        alt={character.name}
      />
      <div className="character-card-name-sub">
        <p className="character-card-name">{character.name}</p>
        <FavoriteButton characterId={character.id} />
      </div>
      <p className="character-card-info-title">Information:</p>
      <div className="character-card-info">
        <div className="character-card-sub">
          <p>
            Status: <span>{character.status}</span>
          </p>
          <p>
            Specie: <span>{character.species}</span>
          </p>
          <p>
            Type: <span>{character.type}</span>
          </p>
          <p>
            Gender: <span>{character.gender}</span>
          </p>
        </div>
        <div className="character-card-sub">
          <p>
            Origin: <span>{character.origin.name}</span>
          </p>
          <p>
            Location: <span>{character.location.name}</span>
          </p>
          <p>
            Appearences: <span>{character.episode.length}</span>
          </p>
        </div>
      </div>
    </div>
  )
}
