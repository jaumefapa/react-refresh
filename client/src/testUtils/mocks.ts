export const MOCK_FAVORITE = {
  id: 181,
  userId: 2,
  characterId: 104,
  createdAt: '2024-02-11T20:25:01.363Z',
  updatedAt: '2024-02-11T20:25:01.363Z',
}

export const SINGLE_CHARACTER = {
  id: 1,
  name: 'Rick Sanchez',
  status: 'Alive',
  species: 'Human',
  type: '',
  gender: 'Male',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  image: 'https://rickandmortyapi.com/api/character/avatar/1.jpeg',
  episode: [
    'https://rickandmortyapi.com/api/episode/1',
    'https://rickandmortyapi.com/api/episode/2',
  ],
  url: 'https://rickandmortyapi.com/api/character/1',
  created: '2017-11-04T18:48:46.250Z',
}

export const ARRAY_OF_CHARACTERS = [
  SINGLE_CHARACTER,
  {
    id: 2,
    name: 'Morty Smith',
    status: 'Alive',
    species: 'Human',
    type: '',
    gender: 'Male',
    origin: {
      name: 'unknown',
      url: '',
    },
    location: {
      name: 'Citadel of Ricks',
      url: 'https://rickandmortyapi.com/api/location/3',
    },
    image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
    episode: [
      'https://rickandmortyapi.com/api/episode/1',
      'https://rickandmortyapi.com/api/episode/2',
      'https://rickandmortyapi.com/api/episode/3',
      'https://rickandmortyapi.com/api/episode/4',
    ],
    url: 'https://rickandmortyapi.com/api/character/2',
    created: '2017-11-04T18:50:21.651Z',
  },
  {
    id: 3,
    name: 'Summer Smith',
    status: 'Alive',
    species: 'Human',
    type: '',
    gender: 'Female',
    origin: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    image: 'https://rickandmortyapi.com/api/character/avatar/3.jpeg',
    episode: ['https://rickandmortyapi.com/api/episode/6'],
    url: 'https://rickandmortyapi.com/api/character/3',
    created: '2017-11-04T19:09:56.428Z',
  },
  {
    id: 4,
    name: 'Beth Smith',
    status: 'Alive',
    species: 'Human',
    type: '',
    gender: 'Female',
    origin: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    location: {
      name: 'Earth (Replacement Dimension)',
      url: 'https://rickandmortyapi.com/api/location/20',
    },
    image: 'https://rickandmortyapi.com/api/character/avatar/4.jpeg',
    episode: [
      'https://rickandmortyapi.com/api/episode/6',
      'https://rickandmortyapi.com/api/episode/7',
    ],
    url: 'https://rickandmortyapi.com/api/character/4',
    created: '2017-11-04T19:22:43.665Z',
  },
]

export const MOCK_REDUX_STATE_CHARACTERS = {
  loading: false,
  success: true,
  error: null,
  charactersInfo: {
    currentPage: 1,
    results: ARRAY_OF_CHARACTERS,
  },
}
