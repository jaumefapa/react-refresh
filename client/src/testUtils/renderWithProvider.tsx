import React, { PropsWithChildren } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import type { AppStore, RootState } from '../redux/store';
import authReducer from '../redux/features/auth/authSlice';
import charactersReducer from '../redux/features/characters/charactersSlice'
import favoritesReducer from '../redux/features/favorites/favoritesSlice'
import { BrowserRouter } from 'react-router-dom'

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: Partial<RootState>
  store?: AppStore
}

const defaultTestingStore = (preloadedState = {}) => {
  return configureStore({
  reducer: {
    auth: authReducer,
    characters: charactersReducer,
    favorites: favoritesReducer,
  },
  preloadedState
})}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    ...renderOptions
  }: ExtendedRenderOptions = {}
  ) {
  const store = defaultTestingStore(preloadedState);
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <BrowserRouter >
          {children}
        </BrowserRouter>
      </Provider>)
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}