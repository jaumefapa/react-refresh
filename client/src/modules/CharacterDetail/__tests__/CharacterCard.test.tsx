import { describe, it, expect } from 'vitest'
import { screen, fireEvent } from '@testing-library/react';

import { renderWithProviders } from '../../../testUtils/renderWithProvider';

import CharacterCard from "../CharacterCard";

const MOCK_CHARACTER = {
  id: 1,
  name: "Rick Sanchez",
  status: "Alive",
  species: "Human",
  type: "",
  gender: "Male",
  origin: {
      name: "Earth (C-137)",
      url: "https://rickandmortyapi.com/api/location/1"
  },
  location: {
      name: "Citadel of Ricks",
      url: "https://rickandmortyapi.com/api/location/3"
  },
  image: "https://rickandmortyapi.com/api/character/avatar/1.jpeg",
  episode: [
      "https://rickandmortyapi.com/api/episode/1",
      "https://rickandmortyapi.com/api/episode/2",
  ],
  url: "https://rickandmortyapi.com/api/character/1",
  created: "2017-11-04T18:48:46.250Z"
}

describe('CharacterCard', () => {
  it('renders the CharacterCard properly', () => {

    const component = renderWithProviders(<CharacterCard character={MOCK_CHARACTER} />)

    expect(component).toMatchSnapshot()
  })
  it("Renders the card with the right content", () =>{
    renderWithProviders(<CharacterCard character={MOCK_CHARACTER} />);

    const name = screen.queryByText(MOCK_CHARACTER.name)
    const status = screen.queryByText(MOCK_CHARACTER.status)
    const species = screen.queryByText(MOCK_CHARACTER.species)
    const gender = screen.queryByText(MOCK_CHARACTER.gender)
    const origin = screen.queryByText(MOCK_CHARACTER.origin.name)
    const location = screen.queryByText(MOCK_CHARACTER.location.name)
    const appearences = screen.queryByText(MOCK_CHARACTER.episode.length)

    expect(name).not.toBeNull();
    expect(status).not.toBeNull();
    expect(species).not.toBeNull();
    expect(gender).not.toBeNull();
    expect(origin).not.toBeNull();
    expect(location).not.toBeNull();
    expect(appearences).not.toBeNull();
  })

  it('The button changes back and forth when clicked', async () => {
    renderWithProviders(
      <CharacterCard  character={MOCK_CHARACTER} />
    )

    const buttonRemoveFav = screen.getAllByLabelText('Remove from favorites');
    expect(buttonRemoveFav).not.toBeNull();
    
    fireEvent.click(screen.getByRole('button'));
    
    const buttonAddFav = screen.getAllByLabelText('Add to favorites');
    expect(buttonAddFav).not.toBeNull();

    fireEvent.click(screen.getByRole('button'));

    expect(buttonRemoveFav).not.toBeNull();
  })
});