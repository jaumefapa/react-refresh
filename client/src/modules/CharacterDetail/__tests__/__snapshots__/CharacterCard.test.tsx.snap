// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CharacterCard > The component UI should not change 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="character-card-container"
      >
        <img
          alt="Rick Sanchez"
          class="character-card-image"
          src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        />
        <div
          class="character-card-name-sub"
        >
          <p
            class="character-card-name"
          >
            Rick Sanchez
          </p>
          <div
            class="favorite-button-container"
          >
            <button
              aria-label="Remove from favorites"
            >
              <svg
                height="1em"
                viewBox="0 0 36 36"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M35.885 11.833c0-5.45-4.418-9.868-9.867-9.868c-3.308 0-6.227 1.633-8.018 4.129c-1.791-2.496-4.71-4.129-8.017-4.129c-5.45 0-9.868 4.417-9.868 9.868c0 .772.098 1.52.266 2.241C1.751 22.587 11.216 31.568 18 34.034c6.783-2.466 16.249-11.447 17.617-19.959c.17-.721.268-1.469.268-2.242"
                  fill="none"
                  stroke="#808080"
                />
              </svg>
            </button>
          </div>
        </div>
        <p
          class="character-card-info-title"
        >
          Information:
        </p>
        <div
          class="character-card-info"
        >
          <div
            class="character-card-sub"
          >
            <p>
              Status: 
              <span>
                Alive
              </span>
            </p>
            <p>
              Specie: 
              <span>
                Human
              </span>
            </p>
            <p>
              Type: 
              <span />
            </p>
            <p>
              Gender: 
              <span>
                Male
              </span>
            </p>
          </div>
          <div
            class="character-card-sub"
          >
            <p>
              Origin: 
              <span>
                Earth (C-137)
              </span>
            </p>
            <p>
              Location: 
              <span>
                Citadel of Ricks
              </span>
            </p>
            <p>
              Appearences: 
              <span>
                2
              </span>
            </p>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="character-card-container"
    >
      <img
        alt="Rick Sanchez"
        class="character-card-image"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <div
        class="character-card-name-sub"
      >
        <p
          class="character-card-name"
        >
          Rick Sanchez
        </p>
        <div
          class="favorite-button-container"
        >
          <button
            aria-label="Remove from favorites"
          >
            <svg
              height="1em"
              viewBox="0 0 36 36"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M35.885 11.833c0-5.45-4.418-9.868-9.867-9.868c-3.308 0-6.227 1.633-8.018 4.129c-1.791-2.496-4.71-4.129-8.017-4.129c-5.45 0-9.868 4.417-9.868 9.868c0 .772.098 1.52.266 2.241C1.751 22.587 11.216 31.568 18 34.034c6.783-2.466 16.249-11.447 17.617-19.959c.17-.721.268-1.469.268-2.242"
                fill="none"
                stroke="#808080"
              />
            </svg>
          </button>
        </div>
      </div>
      <p
        class="character-card-info-title"
      >
        Information:
      </p>
      <div
        class="character-card-info"
      >
        <div
          class="character-card-sub"
        >
          <p>
            Status: 
            <span>
              Alive
            </span>
          </p>
          <p>
            Specie: 
            <span>
              Human
            </span>
          </p>
          <p>
            Type: 
            <span />
          </p>
          <p>
            Gender: 
            <span>
              Male
            </span>
          </p>
        </div>
        <div
          class="character-card-sub"
        >
          <p>
            Origin: 
            <span>
              Earth (C-137)
            </span>
          </p>
          <p>
            Location: 
            <span>
              Citadel of Ricks
            </span>
          </p>
          <p>
            Appearences: 
            <span>
              2
            </span>
          </p>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
